src/packet_analysis/protocol/ip/IPBasedAnalyzer.h

```diff
@@ -0,0 +1,182 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#pragma once
+
+#include "zeek/packet_analysis/Analyzer.h"
+#include "zeek/packet_analysis/Component.h"
+#include "zeek/analyzer/Analyzer.h"
+#include "zeek/analyzer/Manager.h"
+
+namespace zeek::analyzer::pia { class PIA; }
+
+namespace zeek::packet_analysis::IP {
+
+class IPBasedTransportAnalyzer;
+
+/**
+ * A base class for any packet analyzer based on IP. This is used by default by
+ * the TCP, UDP, and ICMP analyzers to reduce a large amount of duplicated code
+ * that those plugins have in common.
+ */
+class IPBasedAnalyzer : public Analyzer {
+public:
+	~IPBasedAnalyzer() override;
+
+	/**
+	 * Returns true if the analyzer determines that in fact a new
+	 * connection has started without the connection statement having
+	 * terminated the previous one, i.e., the new data is arriving at
+	 * what's the analyzer for the previous instance. This is used only
+	 * for TCP.
+	 */
+	virtual bool IsReuse(double t, const u_char* pkt) { return false; }
+
+	/**
+	 * TODO: comment
+	 */
+	virtual void CreateTransportAnalyzer(Connection* conn, IPBasedTransportAnalyzer*& root,
```

Comment:
I moved the code out of `analyzer::Manager` and reworked things a little bit in cb72668d19fb2abfaa6f12353b6cbf0442c092fd. I'm debating whether moving the port registration maps over into the packet analyzers (and splitting them up) makes sense. I think the registration methods can stay but the actual maps can move. `analyzer::Manager` would have to cache pointers to the UDP/TCP analyzers to do the registration/unregistration but that wouldn't be a huge change.

Reply:
yeah, sounds reasonable.

Reply:
Scratch that, I'm going to hold off on this until I do the TCP analyzer so I can move all of it at once. I'm also not sure how I want to store the analyzers there in the manager yet, since I don't really want to add more lookups against `TransportProto` when I'm trying to remove as much of that as I can. That said, all of those registrations/lookups are based on the `PortProto()` value from a `PortVal`, which is a fixed enum anyways.